ext/drx/drx.c

```diff
@@ -2529,17 +2529,24 @@ advance_state(int *detect_state, int new_detect_state, int *skip_unknown_instr_c
 
 /* Advances to state 0 if counter has exceeded threshold, returns otherwise. */
 static inline void
-skip_unknown_instr_inc(int *detect_state, int *skip_unknown_instr_count)
+skip_unknown_instr_inc(int *detect_state, int *skip_unknown_instr_count, int reset_state)
 {
     if (*skip_unknown_instr_count++ >= DRX_RESTORE_EVENT_SKIP_UNKNOWN_INSTR_MAX) {
-        advance_state(detect_state, DRX_DETECT_RESTORE_AVX512_GATHER_EVENT_STATE_0,
-                      skip_unknown_instr_count);
+        advance_state(detect_state, reset_state, skip_unknown_instr_count);
     }
 }
 
 static bool
-drx_try_to_detect_avx512_gather_sequence(void *drcontext, dr_restore_state_info_t *info,
-                                         instr_t *inst, scatter_gather_info_t *sg_info)
+drx_try_to_detect_scatter_gather_sequence(
+    void *drcontext, dr_restore_state_info_t *info, instr_t *inst,
+    scatter_gather_info_t *sg_info,
+    bool (*state_machine_func)(void *drcontext, byte *pc, byte *prev_pc,
+                               int *detect_state, byte **restore_dest_mask_start,
+                               byte **restore_scratch_mask_start,
+                               int *skip_unknown_instr_count, reg_id_t *the_scratch_xmm,
+                               reg_id_t *gpr_bit_mask, reg_id_t *gpr_save_scratch_mask,
+                               uint *scalar_mask_update_no, dr_restore_state_info_t *info,
+                               instr_t *inst, scatter_gather_info_t *sg_info))
```

Comment:
I think most people would agree that this is just too many parameters.  A struct is more readable and manageable.

Reply:
Ok, done.